src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2409,13 +2419,15 @@ public void RecordRecognizedPattern<T> (T accessedItem, Action mark)
 
 			public void RecordUnrecognizedPattern (string message)
 			{
+				Debug.Assert (_patternAnalysisAttempted, "To correctly report all patterns, when starting to analyze a pattern the AnalyzingPattern must be called first.");
+
 				_patternReported = true;
 				_context.ReflectionPatternRecorder.UnrecognizedReflectionAccessPattern (MethodCalling, MethodCalled, message);
 			}
 
 			public void Dispose ()
```

Comment:
What about removing Dispose altogether and add RecordUnrecognizedPattern to places where we are missing them if there are any?

Reply:
I added the `RecordUnrecognizedPattern` in all places, but I kept the Dispose as a way to check that it's everywhere. Still not super clean, but should be reasonably robust now. Added a bunch of tests to cover the interesting cases.